src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,366 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.util;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.Sets;
+import com.google.common.primitives.Ints;
+import org.apache.lucene.index.CheckIndex;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.store.SimpleFSDirectory;
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.ElasticsearchIllegalStateException;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.io.FileSystemUtils;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.ImmutableSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.env.ShardLock;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.settings.IndexSettings;
+import org.elasticsearch.index.shard.*;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ */
+public class MultiDataPathUpgrader {
+
+    private final NodeEnvironment nodeEnvironment;
+    private final ESLogger logger = Loggers.getLogger(getClass());
+
+    public MultiDataPathUpgrader(NodeEnvironment nodeEnvironment) {
+        this.nodeEnvironment = nodeEnvironment;
+    }
+
+    /**
+     * Upgrades the given shard Id from multiple shard paths into the given target path.
+     *
+     * @see #pickShardPath(org.elasticsearch.index.shard.ShardId)
+     */
+    public void upgrade(ShardId shard, ShardPath targetPath) throws IOException {
+        final Path[] paths = nodeEnvironment.shardPaths(shard); // custom data path doesn't need upgrading
+        if (isTargetPathConfigured(paths, targetPath) == false) {
+            throw new IllegalArgumentException("shard path must be one of the shards data paths");
+        }
+        if (paths.length == 1) {
```

Comment:
this seems to be called when `upgrader.needsUpgrading(shardId)` indicates true, so this is not needed. Do we want an assert `needsUgrading(shardId)` instead? if we want the log (which should be debug imho) it should be moved to the calling function. It's not visible now.